lucene/core/src/java/org/apache/lucene/search/QueryVisitor.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.Set;
+
+import org.apache.lucene.index.Term;
+
+/**
+ * Allows recursion through a query tree
+ *
+ * @see Query#visit(QueryVisitor)
+ */
+public abstract class QueryVisitor {
+
+  /**
+   * Called by leaf queries that match on a specific term
+   *
+   * @param query the leaf query
+   * @param term  the term the query will match on
+   */
+  public void consumesTerm(Query query, Term term) { }
+
+  /**
+   * Called by leaf queries that do not match on terms
+   * @param query the query
+   */
+  public void visitLeaf(Query query) { }
```

Comment:
Not knowing all the internals of this and their potential future usages, I am looking at the API mainly and I have a couple of questions:

- should all compound queries have their corresponding `visit` method which calls `getSubVisitor` for each of their sub-query?

- when should `visitLeaf` be used compared to `consumeTerms`? I understand from the javadocs above that there are queries which are not compound queries but don't hold terms either.  If so what would query visitors generally do in their `visitLeaf` impl? I think I am getting to whether `consumesTerm` and `visitLeaf` could be combined together somehow.

Reply:
Yes, all compound queries should pull a sub visitor and call `visit` on each of their sub queries with it.  And `visitLeaf` is for queries that don't match on terms.  Maybe `visitTermLeaf` might be a better name than `consumesTerm`?  In future we might want to add `visitBKDRange` or something similar - for example, to implement range query rewriting like Adrien mentioned on LUCENE-8708

Reply:
> Maybe `visitTermLeaf` might be a better name than `consumesTerm`?

I would rather look for a way to make `visitLeafQuery` clearer, but it may be my own issue with its naming. I find `consumeTerms` clear.